File: shared/src/protocol.ts
import type { ApiError } from './errors.js';
import type { AcpFrame } from './frames.js';
import type { StopReason } from './constants.js';

/**
 * 前后端之间的流式 wire 协议：NDJSON，一行一个 JSON 对象。
 *
 * 为什么不用 SSE：前端无论如何都要手解（EventSource 不支持 POST body，而 prompt
 * 必须 POST），SSE 的 `data:` 转义是纯负担；更关键的是 SSE 的 `id:` /
 * `Last-Event-ID` 语义**暗示可以断点续传**，而 BeginLogOffset 实测是死参数、
 * 服务端根本没有增量续传——用 SSE 等于在协议层撒谎。
 *
 * 帧一律**原样透传**（`body` 就是上游那一帧，不重塑、不改名），
 * 这样 wire 形状与录制 fixture 同形，live 与 mock 才能共用同一个 reducer。
 * 后端唯一加工的地方是错误归一化。
 */
export type WireEvent =
  | WireMeta
  | WireFrame
  | WireHeartbeat
  | WireError
  | WireDone;

/** 流的第一条：告诉前端这一轮的 rid。 */
export interface WireMeta {
  type: 'meta';
  rid: string;
  sessionId: string;
  mock: boolean;
  startedAt: number;
}

export interface WireFrame {
  type: 'frame';
  rid: string;
  /**
   * 上游给的帧序号，可能缺失（归档合并态的帧就没有）。
   * 前端只用于展示与去重，**绝不持久化**：空闲约 5 分钟后服务端计数器会重置。
   */
  offset: number | undefined;
  body: AcpFrame;
}

/** 心跳：防中间层空闲回收连接，同时给前端一个判活信号（存活判定不能挂在 rAF 上）。 */
export interface WireHeartbeat {
  type: 'hb';
  t: number;
}

export interface WireError {
  type: 'error';
  rid: string;
  error: ApiError;
}

/**
 * 流的最后一条。
 *
 * `stopReason` 为空表示上游给了 Result 但原因不认识；而"生成器结束了却从未出现
 * Result.stopReason"不会走到 done，而是发一条 error(kind=stream_break)——
 * 静默截断不能当成功。
 */
export interface WireDone {
  type: 'done';
  rid: string;
  stopReason: StopReason | undefined;
  rawStopReason: string | undefined;
  frameCount: number;
}

export const WIRE_CONTENT_TYPE = 'application/x-ndjson';

/** 解析一行 wire 数据；不是合法对象或 type 不认识时返回 undefined。 */
export function parseWireLine(line: string): WireEvent | undefined {
  const trimmed = line.trim();
  if (!trimmed) return undefined;
  let value: unknown;
  try {
    value = JSON.parse(trimmed);
  } catch {
    return undefined;
  }
  if (typeof value !== 'object' || value === null) return undefined;
  const type = (value as { type?: unknown }).type;
  if (type !== 'meta' && type !== 'frame' && type !== 'hb' && type !== 'error' && type !== 'done') {
    return undefined;
  }
  return value as WireEvent;
}

File: server-python/src/das_python/protocol.py
"""前后端之间的流式 wire 协议：NDJSON，一行一个 JSON 对象。

为什么不用 SSE：前端无论如何都要手解（EventSource 不支持 POST body），SSE 的
`data:` 转义是纯负担；更关键的是 SSE 的 `id:`/`Last-Event-ID` 语义**暗示可以断点续传**，
而 BeginLogOffset 实测是死参数、服务端根本没有增量续传——用 SSE 等于在协议层撒谎。

帧一律**原样透传**（body 就是上游那一帧，不重塑、不改名），这样 wire 形状与录制
fixture 同形，live 与 mock 才能共用同一个 reducer。后端唯一加工的地方是错误归一化。
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, is_dataclass
from typing import Any

WIRE_CONTENT_TYPE = "application/x-ndjson"


@dataclass
class WireMeta:
    """流的第一条：告诉前端这一轮的 rid 与校验码。"""

    rid: str
    sessionId: str
    mock: bool
    startedAt: int
    type: str = "meta"


@dataclass
class WireFrame:
    """帧原样透传：body 就是上游那一帧。offset 可能缺失，绝不持久化。"""

    rid: str
    offset: int | None
    body: dict[str, Any]
    type: str = "frame"


@dataclass
class WireHeartbeat:
    """心跳：防中间层空闲回收连接，同时给前端一个判活信号（存活判定不能挂在 rAF 上）。"""

    t: int
    type: str = "hb"


@dataclass
class WireError:
    rid: str
    error: dict[str, Any]
    type: str = "error"


@dataclass
class WireDone:
    """流最后一条。stopReason 为空表示上游给了 Result 但原因不认识。"""

    rid: str
    stopReason: str | None
    rawStopReason: str | None
    frameCount: int
    type: str = "done"


WireEvent = WireMeta | WireFrame | WireHeartbeat | WireError | WireDone


def serialize(event: WireEvent) -> str:
    """事件 → 一行 NDJSON（不含换行符）。

    与 Node 的 JSON.stringify 字节级对齐：undefined 字段被省略而不是输出 null
    （比如未拿到 offset 的 WireFrame 没有 offset 键、终态未知的 done 没有 stopReason 键）。
    只剥事件顶层的 None；body（上游帧原样透传）里的 null 是上游数据，照常输出。
    """
    raw = asdict(event) if is_dataclass(event) else event
    data = {k: v for k, v in raw.items() if v is not None}
    return json.dumps(data, ensure_ascii=False, separators=(",", ":"))


def parse_wire_line(line: str) -> dict[str, Any] | None:
    """解析一行 wire 数据；不是合法对象或 type 不认识时返回 None。"""
    trimmed = line.strip()
    if not trimmed:
        return None
    try:
        value = json.loads(trimmed)
    except (json.JSONDecodeError, ValueError):
        return None
    if not isinstance(value, dict):
        return None
    if value.get("type") not in ("meta", "frame", "hb", "error", "done"):
        return None
    return value
